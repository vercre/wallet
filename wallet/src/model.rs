//! Model for the wallet application state.

pub mod credential;
mod issuance;

pub use credential::CredentialState;
use issuance::IssuanceState;
use serde::Serialize;

use super::Aspect;
use crate::capabilities::store::{Store, StoreEntry};

/// State for the wallet application.
#[derive(Default, Serialize)]
pub struct Model {
    /// Which aspect of the application is currently active.
    pub active_view: Aspect,

    /// Credential state.
    pub credential: CredentialState,

    /// Issuance state.
    pub issuance: Option<IssuanceState>,

    /// Error state.
    pub error: Option<String>,
}

/// Methods to set the application state based on events.
impl Model {
    /// An error has occurred. Set the error state.
    pub fn error(&mut self, error: String) {
        self.active_view = Aspect::Error;
        self.error = Some(error);
    }

    /// Set up the model with an initial state.
    pub fn ready(&mut self) {
        self.active_view = Aspect::CredentialList;
        self.credential = CredentialState::init();
    }

    /// The user has selected a credential in their wallet to view.
    pub fn select_credential(&mut self, id: String) {
        self.active_view = Aspect::CredentialDetail;
        self.credential.id = Some(id);
    }

    /// The credentials have been retrieved from the wallet's store.
    pub fn credentials_loaded(&mut self, entries: Vec<StoreEntry>) {
        self.active_view = Aspect::CredentialList;
        self.credential.set_credentials(entries);
    }

    /// The user has deleted a credential from their wallet.
    pub fn delete_credential(&mut self) {
        self.credential.id = None;
    }

    /// The user wants to scan an issuance offer QR code.
    pub fn scan_issuance_offer(&mut self) {
        self.active_view = Aspect::IssuanceScan;
        self.issuance = None;
    }

    /// The user has scanned an issuance offer QR code so we can initiate a
    /// pre-authorized issuance flow.
    pub fn issuance_offer<Ev>(
        &mut self, encoded_offer: &str, http: crux_http::Http<Ev>, store: Store<Ev>,
    ) {
        self.active_view = Aspect::IssuanceOffer;
        match IssuanceState::from_offer(encoded_offer, http, store) {
            Ok(issuance_state) => self.issuance = Some(issuance_state),
            Err(e) => self.error(e.to_string()),
        }
    }
}
