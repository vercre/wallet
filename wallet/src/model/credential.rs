//! Credential sub-app state.

use serde::{Deserialize, Serialize};
use vercre_holder::Credential;

/// Application state for the credential sub-app.
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[allow(clippy::module_name_repetitions)]
pub struct CredentialState {
    /// Currently selected credential.
    pub id: Option<String>,

    /// Credentials stored in the wallet.
    pub credentials: Vec<Credential>,
}
