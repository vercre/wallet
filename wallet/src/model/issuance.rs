//! Issuance sub-app state.
use std::collections::HashMap;

use anyhow::bail;
use vercre_holder::credential::ImageData;
use vercre_holder::issuance::{
    Accepted, IssuanceFlow, NotAccepted, PreAuthorized, WithOffer, WithToken, WithoutToken,
};
use vercre_holder::{CredentialConfiguration, CredentialOffer, PreAuthorizedCodeGrant};

use crate::provider::Provider;

/// Configuration and image information for an offered credential.
#[derive(Clone, Debug, Default)]
pub struct OfferedCredential {
    /// Credential configuration.
    pub config: CredentialConfiguration,

    /// Logo image data.
    pub logo: Option<ImageData>,

    /// Background image data.
    pub background: Option<ImageData>,

    /// Credential has been received.
    pub received: bool,

    /// Credential has been stored.
    pub stored: bool,
}

impl OfferedCredential {
    /// Determine if the credential logo needs to be fetched.
    pub fn needs_logo(&self) -> bool {
        if self.logo.is_some() {
            return false;
        }
        if let Some(display) = &self.config.display {
            if let Some(logo) = &display[0].logo {
                if logo.uri.is_some() {
                    return true;
                }
            }
        }
        false
    }

    /// Determine if the credential background needs to be fetched.
    pub fn needs_background(&self) -> bool {
        if self.background.is_some() {
            return false;
        }
        if let Some(display) = &self.config.display {
            if let Some(background) = &display[0].background_image {
                if background.uri.is_some() {
                    return true;
                }
            }
        }
        false
    }
}

/// Application state for the issuance sub-app.
#[derive(Clone, Debug, Default)]
#[allow(clippy::module_name_repetitions)]
#[allow(clippy::large_enum_variant)]
pub enum IssuanceState {
    /// No issuance is in progress.
    #[default]
    Inactive,

    /// An offer has been received
    Offered { offer: CredentialOffer, grant: PreAuthorizedCodeGrant },

    /// Issuer metadata has been received. Can use this state to keep updating
    /// the offered credentials' logo and background images.
    IssuerMetadata {
        flow: IssuanceFlow<WithOffer, PreAuthorized, NotAccepted, WithoutToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// The offer has been accepted by the user. Can use this state to update
    /// the PIN number if needed.
    Accepted {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithoutToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// An access token has been received.
    Token {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithToken>,
        offerred: HashMap<String, OfferedCredential>,
    },

    /// A proof has been created. Can use this state to receive credentials and
    /// update the offered list to keep track of outstanding credentials. Can
    /// also use it to keep track of the credentials stored.
    Proof {
        flow: IssuanceFlow<WithOffer, PreAuthorized, Accepted, WithToken>,
        offerred: HashMap<String, OfferedCredential>,
        proof: String,
    },
}

/// State change implementation.
impl IssuanceState {
    /// Create an issuance state from a URL-encoded offer.
    pub fn from_offer(encoded_offer: &str) -> anyhow::Result<Self> {
        let offer_str = urlencoding::decode(encoded_offer)?;
        let offer = serde_json::from_str::<CredentialOffer>(&offer_str)?;

        // Check the offer has a pre-authorized grant. This is the only flow
        // type supported by this wallet (for now).
        let Some(pre_auth_code_grant) = offer.pre_authorized_code() else {
            bail!("grant other than pre-authorized code is not supported");
        };

        Ok(Self::Offered {
            offer,
            grant: pre_auth_code_grant,
        })
    }

    /// Cancel the issuance process.
    pub fn cancel<Ev>(&mut self, _provider: &Provider<Ev>) -> anyhow::Result<()>
    where
        Ev: 'static,
    {
        // TODO: Reset state
        Ok(())
    }
}
