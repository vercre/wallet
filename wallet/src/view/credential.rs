//! Credential view models

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use vercre_holder::credential::{Credential as CredentialModel, ImageData};
use vercre_holder::Claim;

use crate::model::CredentialState;

/// View model for a verifiable credential.
///
/// Matches the `Credential` model from the `vercre-holder` crate as closely as
/// possible but caters for ease of code generation for shells.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Credential {
    /// Credential `id` is the credential's unique identifier
    /// (from Verifiable Credential `id` or generated if credential has no
    /// `id`).
    pub id: String,

    /// The credential issuer.
    pub issuer: String,

    /// The Verifiable Credential as issued, for use in Presentation
    /// Submissions. This could be a base64-encoded JWT or 'stringified'
    /// JSON.
    pub issued: String,

    /// The credential type. Used to determine whether a credential matches a
    /// presentation request.
    #[serde(rename = "type")]
    pub type_: Vec<String>,

    /// Credential format. Information on how the encoded credential is
    /// formatted.
    pub format: String,

    /// Claims display. The key is the subject ID and the value is the claims
    /// displayed as a string with a claim on each line and nested claims
    /// indented.
    pub claims: HashMap<String, String>,

    /// The date the credential was issued as an RFC3339 string.
    pub issuance_date: String,

    /// The date the credential is valid from as an RFC3339 string. Empty string
    /// if not applicable.
    pub valid_from: String,

    /// The date the credential is valid until (expiry) as an RFC3339 string.
    /// Empty string if not applicable.
    pub valid_until: String,

    /// Name from issuer's metadata.
    pub name: String,

    /// Description from issuer's metadata.
    pub description: String,

    /// Background color from issuer's metadata. Empty string if not applicable.
    pub background_color: String,

    /// A base64-encoded logo image for the credential ingested from the logo
    /// url in the display section of the metadata.
    ///
    /// The elements of `ImageData` will be empty strings if the logo has not
    /// been set.
    pub logo: ImageData,

    /// A base64-encoded background image for the credential ingested from the
    /// url in the display section of the metadata.
    ///
    /// The elements of `ImageData` will be empty strings if the background has
    /// not been set.
    pub background: ImageData,
}

impl From<CredentialModel> for Credential {
    fn from(credential: CredentialModel) -> Self {
        let mut claims = HashMap::new();
        for claim_set in &credential.subject_claims {
            let mut display = String::new();
            claims_display(
                &mut display,
                claim_set.claims.clone(),
                credential.claim_definitions.clone().unwrap_or_default(),
                0,
            );
            claims.insert(claim_set.id.clone().unwrap_or_default(), display);
        }
        let (name, description, background_color) = match credential.display {
            Some(display_list) => {
                // TODO: Locale support.
                let display = display_list
                    .clone()
                    .into_iter()
                    .find(|display| display.locale.is_none())
                    .unwrap_or_else(|| display_list[0].clone());
                (
                    display.name,
                    display.description.unwrap_or_default(),
                    display.background_color.unwrap_or_default(),
                )
            }
            None => (String::new(), String::new(), String::new()),
        };
        Self {
            id: credential.id,
            issuer: credential.issuer,
            issued: credential.issued,
            type_: credential.type_,
            format: credential.format,
            claims: HashMap::new(), // TODO: Implement claims_display
            issuance_date: credential.issuance_date.to_rfc3339(),
            valid_from: credential.valid_from.map_or_else(String::new, |date| date.to_rfc3339()),
            valid_until: credential.valid_until.map_or_else(String::new, |date| date.to_rfc3339()),
            name,
            description,
            background_color,
            logo: credential.logo.unwrap_or_default(),
            background: credential.background.unwrap_or_default(),
        }
    }
}

/// View for the verifiable credential sub-app
#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct CredentialView {
    /// Currently active credential ID
    pub id: Option<String>,

    /// List of stored credentials
    pub credentials: Vec<Credential>,
}

impl From<CredentialState> for CredentialView {
    fn from(state: CredentialState) -> Self {
        Self {
            id: state.id,
            credentials: state.credentials.into_iter().map(Credential::from).collect(),
        }
    }
}

// Convert the `SubjectClaims` model to a string representation, including
// using the claim definitions to find the claim labels.
// TODO: Use locales.
#[allow(dead_code)]
fn claims_display(
    display: &mut String, claims: Map<String, Value>, definitions: HashMap<String, Claim>,
    indent_level: usize,
) {
    let indent = "  ".repeat(indent_level);
    for (key, value) in claims {
        // Find a label if possible, otherwise use the key.
        if let Some(definition) = definitions.get(&key) {
            match definition {
                Claim::Entry(def) => {
                    if let Some(def_display) = def.display.clone() {
                        let locale_display = def_display
                            .clone()
                            .into_iter()
                            .find(|d| d.locale.is_none())
                            .unwrap_or_else(|| def_display[0].clone());
                        display.push_str(&format!("{}{}: ", indent, locale_display.name));
                    } else {
                        let label = capitalize(&key);
                        display.push_str(&format!("{}{}: ", indent, label));
                    };
                }
                Claim::Set(_) => {
                    let label = capitalize(&key);
                    display.push_str(&format!("{}{}: ", indent, label));
                }
            }
        } else {
            let label = capitalize(&key);
            display.push_str(&format!("{}{}: ", indent, label));
        }
        // And now the value. Might need to recurse if it's a nested claim.
        match value {
            Value::Object(map) => {
                display.push('\n');
                claims_display(display, map, definitions.clone(), indent_level + 1);
            }
            _ => {
                display.push_str(&format!("{}\n", value));
            }
        }
    }
    *display = display.replace("\"", "");
}

// Capitalize the first letter of a string.
pub fn capitalize(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        None => String::new(),
        Some(c) => c.to_uppercase().collect::<String>() + chars.as_str(),
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn test_claims_display() {
        let json = include_bytes!("../model/credentials.json");
        let credentials: Vec<CredentialModel> =
            serde_json::from_slice(json).expect("should deserialize");
        assert_eq!(credentials.len(), 2);

        let claim_sets = credentials[0].subject_claims.clone();
        assert_eq!(claim_sets.len(), 1);
        let claims = claim_sets[0].claims.clone();

        let Some(defs) = credentials[0].claim_definitions.clone() else {
            panic!("No claim definitions found");
        };

        let display = &mut String::new();
        claims_display(display, claims, defs, 0);
        assert_yaml_snapshot!(display);
    }
}
