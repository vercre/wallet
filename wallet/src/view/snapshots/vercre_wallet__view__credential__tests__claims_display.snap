---
source: wallet/src/view/credential.rs
assertion_line: 219
expression: display
snapshot_kind: text
---
"Address: \n  Locality: Wellington\n  Street_address: 123 Fake St\nEmail: normal.user@example.com\nFamily name: Person\nGiven name: Normal\n"
